### PR TITLE
feat: handwritten-proof AndOrXor

### PR DESCRIPTION
**Pull request by tobiasgrosser**
This one is again pulled out from https://github.com/opencompl/ssa/pull/167.

### CODE REVIEW

File: SSA/Projects/InstCombine/AliveHandwrittenLargeExamples.lean [hunk 1]
```diff
@@ -6,6 +7,16 @@ open MLIR AST
...
+namespace LLVMTheory
+
+@[simp]
+theorem LLVM.const?_eq : LLVM.const? i = .some (BitVec.ofInt w i) := rfl
+
+@[simp]
+theorem LLVM.xor?_eq : LLVM.xor? a b  = .some (BitVec.xor a b) := rfl
+
+end LLVMTheory
+
```
Comment (bollu) [inline SSA/Projects/InstCombine/AliveHandwrittenLargeExamples.lean:19]: Let's move these to where `LLVM.const?` and friends are defined, at https://github.com/opencompl/ssa/blob/efac4d7ad48716e345b2e52d48fc8b345333f4d7/SSA/Projects/InstCombine/LLVM/Semantics.lean#L411

File: SSA/Projects/InstCombine/AliveHandwrittenLargeExamples.lean [hunk 2]
```diff
@@ -50,4 +61,102 @@ axioms: [propext, Classical.choice, Quot.sound] -/
...
+  rcases c3 with none | c3 <;>
+  rcases x with none | x <;>
+  simp only [LLVM.xor?_eq, xor_eq, Option.bind_eq_bind, Option.none_bind, Option.bind_none,
+    Option.some_bind, Refinement.refl]
+  rw [←Option.bind_eq_bind]
+  simp_alive_ops
+  by_cases h : w ≤ BitVec.toNat c2 <;>
+  simp only [ge_iff_le, h, ↓reduceIte, Option.bind_eq_bind, Option.none_bind, Option.bind_none,
+    Refinement.refl, Option.some_bind, h, Option.pure_def, Option.some_bind, Refinement.some_some]
+  simp only [ushr_xor_right_distrib, xor_assoc]
```
Comment (bollu) [inline SSA/Projects/InstCombine/AliveHandwrittenLargeExamples.lean:134]: Could have sworn you had a cleaner proof @tobiasgrosser :)

File: SSA/Projects/InstCombine/AliveHandwrittenLargeExamples.lean [hunk 3]
```diff
@@ -50,4 +61,102 @@ axioms: [propext, Classical.choice, Quot.sound] -/
...
+
+#print axioms alive_simplifyAndOrXor2515
```
Comment (bollu) [inline SSA/Projects/InstCombine/AliveHandwrittenLargeExamples.lean:136]: `#print -> #guard`, as we did last time.

File: SSA/Projects/InstCombine/AliveHandwrittenLargeExamples.lean (hunk 1 above)
Comment (tobiasgrosser) [reply SSA/Projects/InstCombine/AliveHandwrittenLargeExamples.lean]: Done.

File: SSA/Projects/InstCombine/AliveHandwrittenLargeExamples.lean (hunk 2 above)
Comment (tobiasgrosser) [reply SSA/Projects/InstCombine/AliveHandwrittenLargeExamples.lean]: A managed to clean it up a bit, but not by much.

**Comment by bollu** [comment]
LGTM, I vaguely remember there being a nicer way to write:

```lean
rcases c1 with rfl | c1 <;> try (simp; done)
```

That is, the trivial case should be dealt with by ` simp`, and the complex case with `.some` should not have the proof state be changed. Ping @alexkeizer / @goens for some post-merge cleanup!

**Pull request merged**